### PR TITLE
don't move keys if the key where already moved in a previous migration run

### LINKED ISSUE #17564: files_encryption folder structure after migration to 8.1
After running the occ encryption key migration I noticed that for one user the filekeys were migrated to `data/<user>/files_encryption/keys/files/files/files/<path>/<filename>/OC_DEFAULT_MODULE/OC_DEFAULT_MODULE/OC_DEFAULT_MODULE/fileKey` (similar for the sharekeys). Not sure what triggered this nesting, but it's definitely not correct.
For the other users the keys were correctly moved to `data/<user>/files_encryption/keys/files/<path>/<filename>/OC_DEFAULT_MODULE/fileKey`. Note that for these users there was also an (empty?) folder structure created at `data/<user>/files_encryption/keys/<path>/<filename>`; these folders do not seem to serve any purpose and can be removed?

### PR DESCRIPTION
**Pull request by schiessle**
fix #17564

steps to test:
- setup ownCloud 8.0
- enable encyption
- migrate to this branch
- run multiple times `./occ encryption:migrate`

without this patch keys get moved each time which crates a nested structure of files/files/files/...
with this patch keys get only moved once

### CODE REVIEW

**Comment by schiesbn** [comment]
@jknockaert can you give it a try? This should solve your issue.

cc @DeepDiver1975 @karlitschek would be nice to get it in 8.1.1

@PVince81 maybe you have time for one more review? Thanks!

**Comment by karlitschek** [comment]
yes. please backport :+1:

**Comment by PVince81** [comment]
Tested, works :+1:

**Comment by jknockaert** [comment]
I already moved everything manually so there's not much left here to be tested.
I had a quick look at the code. It seems quite a lot of lines of code for merely checking if the file-to-be-moved is still there where it is to be moved from, but I guess the approach implemented will do.

**Comment by schiesbn** [comment]
@jknockaert if possible you could set up a test instance from git and then migrate from the stable8 branch to this one, run the migration multiple times and check if all problems you observed in the past are solved. Alternatively you could also do a code review, since @PVince81 and I already tested it. Both would be really appreciated. Thanks a lot!

**Comment by schiesbn** [comment]
test passed successfully here https://github.com/owncloud/core/pull/17716

Can be merged after a second review

**Comment by jknockaert** [comment]
I did a reading of the code. The fix is mainly in L338–441 of `apps/encryption/lib/migration.php`, where a test is added to check if for the encryption key target location a corresponding encrypted file exists. I guess that will do for the bugs we've observed. Some other changes such as L237–239 probably improve general robustness but I'm not aware of any bug being fixed here. Other than that there's a remarkable amount of code added to allow for an error being written to the log. We may want the corresponding log entry to be a bit more informative, now it merely states a filename twice whereas we may want to output the path and a bit more precise textual formulation of the error.

File: apps/encryption/lib/migration.php [hunk 1]
```diff
@@ -270,8 +283,15 @@ private function renameFileKeys($user, $path, $trash = false) {
 						$this->renameFileKeys($user, $path . '/' . $file, $trash);
 					} else {
 						$target = $this->getTargetDir($user, $path, $file, $trash);
-						$this->createPathForKeys(dirname($target));
-						$this->view->rename($user . '/' . $path . '/' . $file, $target);
+						if ($target) {
```
Comment (nickvergessen) [inline apps/encryption/lib/migration.php]: !== false, so target `'0'` works?

**Comment by schiesbn** [comment]
> Some other changes such as L237–239 probably improve general robustness but I'm not aware of any bug being fixed here.

yes, it increases robustness and fixed some issues I discovered during testing, e.g. that all already moved key was destroyed by trying to move a non-existing key again to the target location. All this additional stuff was discovered and fixed thanks to unit tests. Thanks a lot for your code review!

Successfully reviewed/tested by me, @PVince81 and @jknockaert. Backport was already merged so let's get this in too.

File: apps/encryption/lib/migration.php [hunk 2]
```diff
@@ -280,22 +300,48 @@ private function renameFileKeys($user, $path, $trash = false) {
 	}
 
 	/**
+	 * get system mount points
+	 * wrap static method so that it can be mocked for testing
```
Comment (nickvergessen) [inline apps/encryption/lib/migration.php]: mark as internal?

File: apps/encryption/lib/migration.php [hunk 3]
```diff
@@ -280,22 +300,48 @@ private function renameFileKeys($user, $path, $trash = false) {
...
+			$targetDir = $user . '/files_encryption/keys/files/' . $filePath . '/' . $this->moduleId . '/' . $filename;
 		}
 
-		return $targetDir;
+		if ($user === '') {
+			// for system wide mounts we need to check if the mount point really exists
+			$normalized = trim($filePath, '/');
+			$systemMountPoints = $this->getSystemMountPoints();
+			foreach ($systemMountPoints as $mountPoint) {
+				if (strpos($normalized, $mountPoint['mountpoint']) === 0)
```
Comment (nickvergessen) [inline apps/encryption/lib/migration.php]: you removed `/`

so `test` will match the mountpoint `test2`.
Should add the slash at the end for the strpos comparison

Comment (nickvergessen) [inline apps/encryption/lib/migration.php]: This is rather fatal and needs a patch + backport as well

**Comment by schiesbn** [comment]
good catch @nickvergessen  I will prepare a follow-up PR

**Pull request merged**